### ISSUE TITLE
Run image generation and upscaling in parallel

### ISSUE THREAD

**Issue by whosawhatsis** [invoke-ai/InvokeAI#862]
When generating large sets of images, my GPU is pegged while generating images (txt2img), but my CPU is idle. When that finishes, it starts running Real-ESRGAN and GFPGAN on all of those images, which uses close to half of my CPU time and none of my GPU for nearly as long.

This is on an M1 Max, and the optimum might be different when using CUDA, but it would be nice to have an option to run the upscaling for one image in parallel with generating the next image, rather than doing all of the generation, then all of the upscaling after that finishes.

A related issue is that, if I decide to stop the image generation early, nothing gets upscaled. This would also be fixed by my suggestion above. Even if it instead alternated, generating one image, then upscaling it, then generating the next image, and so on, ending generation early could still result in upscaled images. This would result in alternating between CPU and GPU tasks, which might if nothing else, at least have the benefit of reducing the maximum temperature (and thus fan noise level) of the computer during the whole process.

**Comment by psychedelicious** [comment]
I'll note that in the web UI, the sequence is as you request - generation1, upscale1, facefix1, generation2, upscale2, facefix2, and so on. 

If you cancel before an image has gone through request postprocessing, though, it is not saved. I suppose it should save it if you cancel after generation but before post?

**Comment by whosawhatsis** [comment]
Are you sure? I just got the webui running and tried it. I set it to generate 10 images. It did that, then began upscaling them. Maybe you're running an older version or a different branch? I just installed mine two days ago...

**Comment by psychedelicious** [comment]
I suppose you are on the `main` branch? On `development`, there is a brand new UI which works as I described (that's how it should work, anyways).

**Comment by whosawhatsis** [comment]
Ah, cool, I'll check that out.

**Comment by whosawhatsis** [comment]
Yes, the development version of the webui does do the upscaling for each image, which I like a lot better. I still think it would be useful to be able to do that in parallel with generating the next image, though, since (at least on apple silicon), one uses the CPU and the other uses the GPU, so they don't step on each other much.

**Comment by psychedelicious** [comment]
@Kyle0654 With the new backend, will the workers process jobs in parallel e.g. iteration 1's postprocessing can happen in parallel with interation 2's generation?

**Comment by Kyle0654** [comment]
> @Kyle0654 With the new backend, will the workers process jobs in parallel e.g. iteration 1's postprocessing can happen in parallel with interation 2's generation?

It could be done, but it's a challenging problem with the new node-based stuff, since I can't readily tell what resources would need to be locked to process a node, and those could change based on settings.
